src/Ledger/Gov.lagda

```diff
@@ -39,6 +54,64 @@ _⊢_⇀⦇_,GOV⦈_     : GovEnv → GovState → List (GovVote ⊎ GovProposal
+getAidPairsSet : GovState → ℙ (GovActionID × GovActionID)
+getAidPairsSet aid×states =
+  mapPartial (λ (aid , as) → (aid ,_) <$> getHash (prevAction as)) $ fromList aid×states
+
+-- a list of GovActionID pairs connects the first GovActionID to the second
+_connects_to_ : List (GovActionID × GovActionID) → GovActionID → GovActionID → Set
+[] connects aidNew to aidOld = aidNew ≡ aidOld
+((aid , aidPrev) ∷ s) connects aidNew to aidOld = aid ≡ aidNew × s connects aidPrev to aidOld
+                                                  ⊎ s connects aidNew to aidOld
+-- TODO: delete these notes (and possibly add them to docs) after confirming they're correct.
+--   Suppose:  aid₁ → aid₂ → aid₃ → aid₄ → aid₅.
+--   And let:  l := (aid₅, aid₄) (aid₄, aid₃) (aid₃, aid₂) (aid₂, aid₁)
+--   Check:    l connects aid₅ to aid₂ :
+--               ((aid₅, aid₄) ∷ s) connects aid₅ to aid₂ = aid₅ ≡ aid₅ × s connects aid₄ to aid₂
+
+[_connects_to_?] : ∀ l aidNew aidOld → Dec (l connects aidNew to aidOld)
+[ [] connects aidNew to aidOld ?] = aidNew ≟ aidOld
+
+[ (aid , aidPrev) ∷ s connects aidNew to aidOld ?] =
+  ((aid ≟ aidNew) ×-dec [ s connects aidPrev to aidOld ?]) ⊎-dec [ s connects aidNew to aidOld ?]
+
+enactable : EnactState → List (GovActionID × GovActionID) → GovActionID × GovActionState → Set
+enactable e aidPairs = λ (aidNew , as) → case getHashES e (GovActionState.action as) of λ where
+  nothing → ⊤
+  (just aidOld) → ∃[ t ](fromList t ⊆ fromList aidPairs × Unique t × t connects aidNew to aidOld)
+
+open Equivalence
+
+any?-connecting-subperm : ∀ {u}{v} → ∀ L → Dec (Any(λ l → Unique l × l connects u to v) (subpermutations L))
+any?-connecting-subperm {u}{v} L = any? (λ l → unique? _≟_ l ×-dec [ l connects u to v ?]) (subpermutations L)
+
+∃?-connecting-subperm : ∀ {u}{v} → ∀ L → Dec (∃[ l ] l ∈ˡ subpermutations L × Unique l × l connects u to v)
+∃?-connecting-subperm L = from (map′⇔ (↔⇒ Any↔)) (any?-connecting-subperm L)
+
+∃?-connecting-subset : ∀ {u}{v} → ∀ L → Dec (∃[ l ](l ⊆ˡ L × Unique l × l connects u to v))
+∃?-connecting-subset L = from (map′⇔ ∃uniqueSubset⇔∃uniqueSubperm) (∃?-connecting-subperm L)
+
+enactable? : ∀ eState aidPairs aidNew×st → Dec(enactable eState aidPairs aidNew×st)
+enactable? eState aidPairs (aidNew , as) with getHashES eState (GovActionState.action as)
+... | nothing = yes tt
+... | just aidOld = from (∃?-sublist-⇔ th) (∃?-connecting-subset aidPairs)
+
+allEnactable : EnactState → GovState → Set
+allEnactable e aid×states = All (λ p → enactable e (getAidPairsList aid×states) p) aid×states
+
+allEnactable? : (eState : EnactState)(aid×states : GovState) → Dec (allEnactable eState aid×states)
+allEnactable? eState aid×states =
+  all? (λ aid×st → enactable? eState (getAidPairsList aid×states) aid×st) aid×states
+
+maxAllEnactable : EnactState → List (GovActionID × GovActionState) → List (List (GovActionID × GovActionState))
+maxAllEnactable e = maxsublists⊧P{P = λ l → allEnactable e l}{λ l → allEnactable? e l}
```

Comment:
@WhatisRT this function is meant to return all sublists of the given list of proposals that (a) are all enactable and (b) are of maximal length among those sublists that satisfy `allEnactable`.

Next I'd like to do the following:

1. check whether you think this or something similar the right approach toward having a function that removes non-enactable proposals
2. assuming a positive answer to 1, test `maxAllEnactable` with some toy examples of lists of proposals; if you have suggestions or guidance for this step, please advise, otherwise, I'll try to make something up.

Reply:
Looks fine. Can you prove that the chosen list satisfies the property & no larger intermediate list does?

Reply:
@WhatisRT My latest commit contains proofs verifying the assertions made in the comments of `maxAllEnactable`---namely, `maxAllEnactable` returns a list `ls` of sublists of the given `aid×states` such that
1. each sublist `l ∈ ls` satisfies `allEnactable e l` and
2. each sublist `l ∈ ls` is of maximal length among sublists of `aid×states` satisfying `allEnactable`.

Your comment above refers to "the chosen list."  In fact, I didn't defined it so that there is one particular "special" maximal list (in case there is more than one), but if we do want to chose one, then we could simply take, say, the head of the list of maximal-length, `allEnactable` lists returned by `maxAllEnactable`.